internal/throttling/memory_gcra.go

```diff
@@ -0,0 +1,63 @@
+package throttling
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/throttled/throttled/v2"
+	"github.com/throttled/throttled/v2/store/memstore"
+
+	"github.com/rudderlabs/rudder-server/internal/cachettl"
+)
+
+const defaultMaxCASAttemptsLimit = 100
+
+type gcra struct {
+	mu    sync.Mutex
+	store *cachettl.Cache
```

Comment:
Not sure I understand why we need a cache with ttl for storing independent rate limiters per key, when the gcra limiter can work for multiple keys already

Reply:
The underlying GCRA limiter can work with multiple keys but you can't set a different quota per key so I had to create a new limiter each time. Meaning that with the `throttled` GCRA limiter alone we'll have to use the same quota for all destinations.

An optimization I can do is to create one limiter per quota instead of one per key, so if more than one key uses the same quota I don't create an instance for each but that would increase the expiration complexity for little advantage.

In any case, to give a bit more context, I created the `cachettl.Cache` because I did not want something to live in memory forever even if it was used only once (e.g. you get only one message a day/week/whatever for a destination, still keep the limiter in the map forever). I searched a bit for existing implementations because I didn't really want to spend time writing my own cache but I could only find caches where the TTL is set globally and not per-key. It was simple enough to convince me to write it so here we are :-)

Reply:
Each router is for a specific destination type and each destination type has its own quota configuration.
Additionally, this quota configuration doesn't appear reloadable for the time being, i.e. it cannot change.
Based on the above, a single GCRA limiter would suffice. The only place I would justify using a TTL cache and multiple limiters would be in the gorate implementation, where we need to have a different limiter per key.

Reply:
This isn't just about routers. The library lives in `internal/throttling` and it is meant to be used across the board for all throttling needs. In a following PR I'm planning to replace the limiter that we're using on the gateway as well.

That said, what I'm trying to do here is to provide a simple enough interface to throttle with more than one algorithm supported. You don't have to care about which algorithm you're using, you just setup the throttler and call `.Limit()`. The throttler provides different quotas per key and so that applies to all algos, memory GCRA included. I don't think extracting the memory GCRA one makes sense just because of the TTL cache (which in my opinion is a simple enough and lightweight data structure by the way).

After all the benchmarking and the tests I don't see a risk with going live with this to be honest. After some time we can still gather some stats and decide whether we want to review this, but so far I don't see a reason to remove the TTL cache. Of course if you're concerned that this would create issues please elaborate.